# 氢原子光谱实验报告

# 一、 预习报告
![预习报告](预习报告原始数据/氢原子光谱_1.jpg)
![预习报告](预习报告原始数据/氢原子光谱_2.jpg)

# 二、 原始数据
![原始数据](预习报告原始数据/氢原子光谱_3.jpg)

# 三、 实验报告

In [1]:
import pandas as pd
from pint import UnitRegistry
ureg=UnitRegistry()
quan=ureg.Quantity
import numpy as np
import phyexp as pe
from uncertainties import ufloat, umath
pd.set_option('display.float_format', '{:.4f}'.format)  # 固定显示 4 位小数
ureg.default_format = '.4f'   # 设置所有 Quantity 的默认格式

C:\Users\Mcsof\AppData\Local\Temp\ipykernel_28092\1033943161.py:9: DeprecationWarning: This function will be removed in future versions of pint.
Use ureg.formatter.default_format
  ureg.default_format = '.4f'   # 设置所有 Quantity 的默认格式


In [2]:
钠灯角度=pd.DataFrame({
    "k=-2 左": [ureg("346deg+0arcmin"), ureg("315deg+35arcmin"), ureg("289deg+14arcmin"), ureg("251deg+25arcmin"), ureg("220deg+2arcmin")],
    "k=-2 右": [ureg("165deg+59arcmin"), ureg("135deg+30arcmin"), ureg("109deg+7arcmin"), ureg("71deg+22arcmin"), ureg("40deg+0arcmin")],
    "k=+2 左": [ureg("304deg+37arcmin"), ureg("274deg+9arcmin"), ureg("247deg+50arcmin"), ureg("210deg+1arcmin"), ureg("178deg+26arcmin")], 
    "k=+2 右": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([124,94,67,30,358],[32,3,45,0,37])]
})

In [3]:
钠灯角度

,k=-2 左,k=-2 右,k=+2 左,k=+2 右
0,346.0000 degree,165.9833 degree,304.6167 degree,124.5333 degree
1,315.5833 degree,135.5000 degree,274.1500 degree,94.0500 degree
2,289.2333 degree,109.1167 degree,247.8333 degree,67.7500 degree
3,251.4167 degree,71.3667 degree,210.0167 degree,30.0000 degree
4,220.0333 degree,40.0000 degree,178.4333 degree,358.6167 degree


In [4]:
氢灯角度=pd.DataFrame({
    "k=-1 左": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([210, 207, 206],[42, 42, 31])],
    "k=-1 右": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([30, 27, 26],[40, 40, 31])],
    "k=+1 左": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([187, 190, 191],[39, 38, 34])], 
    "k=+1 右": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([7, 10, 11],[39, 37, 34])], 
    "k=-2 左": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([222, 216, np.nan],[31, 17, np.nan])],
    "k=-2 右": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([42, 36, np.nan],[32, 14, np.nan])],
    "k=+2 左": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([175, 182, np.nan],[51, 3, np.nan])], 
    "k=+2 右": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([355, 2, np.nan],[51, 4, np.nan])], 
    "k=-3 左": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([235, 225, np.nan],[37, 22, np.nan])],
    "k=-3 右": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([55, 45, np.nan],[35, 19, np.nan])],
    "k=+3 左": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([162, 173, np.nan],[48, 5, np.nan])], 
    "k=+3 右": [quan(i[0],"deg")+quan(i[1],"arcmin") for i in zip([342, 353, np.nan],[48, 5, np.nan])], 
}, index=["红","蓝","紫"])

In [5]:
氢灯角度

,k=-1 左,k=-1 右,k=+1 左,k=+1 右,k=-2 左,k=-2 右,k=+2 左,k=+2 右,k=-3 左,k=-3 右,k=+3 左,k=+3 右
红,210.7000 degree,30.6667 degree,187.6500 degree,7.6500 degree,222.5167 degree,42.5333 degree,175.8500 degree,355.8500 degree,235.6167 degree,55.5833 degree,162.8000 degree,342.8000 degree
蓝,207.7000 degree,27.6667 degree,190.6333 degree,10.6167 degree,216.2833 degree,36.2333 degree,182.0500 degree,2.0667 degree,225.3667 degree,45.3167 degree,173.0833 degree,353.0833 degree
紫,206.5167 degree,26.5167 degree,191.5667 degree,11.5667 degree,nan degree,nan degree,nan degree,nan degree,nan degree,nan degree,nan degree,nan degree


## 1. 用钠黄光 λ = 589.3 nm 作为标准谱线校准光栅常数 d，并计算不确定度 u(d)。注意消除正负级不严格对称的系统误差。

正负级不严格对称的系统误差的消除：

通过正负级做差除二作为折射角测量结果消除系统误差

In [6]:
钠灯角度["k=±2差值 左"]=钠灯角度["k=-2 左"] - 钠灯角度["k=+2 左"]
钠灯角度["k=±2差值 右"]=钠灯角度["k=-2 右"] - 钠灯角度["k=+2 右"]
钠灯角度["k=±2差值"]=(钠灯角度["k=±2差值 右"]+钠灯角度["k=±2差值 左"])/2
钠灯角度["k=±2差值弧度"]=钠灯角度["k=±2差值"].apply(lambda x: x.to('rad').magnitude)  # 转为弧度数值
钠灯角度["k=±2差值弧度"]=钠灯角度["k=±2差值弧度"].apply(lambda x: x+np.pi if x < 0 else x)  # 调整到0-pi区间
钠灯角度["折射角"]=钠灯角度["k=±2差值弧度"]/2

In [7]:
钠灯角度

,k=-2 左,k=-2 右,k=+2 左,k=+2 右,k=±2差值 左,k=±2差值 右,k=±2差值,k=±2差值弧度,折射角
0,346.0000 degree,165.9833 degree,304.6167 degree,124.5333 degree,41.3833 degree,41.4500 degree,41.4167 degree,0.7229,0.3614
1,315.5833 degree,135.5000 degree,274.1500 degree,94.0500 degree,41.4333 degree,41.4500 degree,41.4417 degree,0.7233,0.3616
2,289.2333 degree,109.1167 degree,247.8333 degree,67.7500 degree,41.4000 degree,41.3667 degree,41.3833 degree,0.7223,0.3611
3,251.4167 degree,71.3667 degree,210.0167 degree,30.0000 degree,41.4000 degree,41.3667 degree,41.3833 degree,0.7223,0.3611
4,220.0333 degree,40.0000 degree,178.4333 degree,358.6167 degree,41.6000 degree,-318.6167 degree,-138.5083 degree,0.7242,0.3621


*最后两列被转换为弧度制，下面的计算也按弧度制进行*

In [8]:
theta=钠灯角度["折射角"].mean()

In [9]:
u_atheta=钠灯角度["折射角"].std()

In [10]:
仪器误差限=分度值=ureg("1arcmin").to("rad").magnitude
u_btheta=仪器误差限/3**0.5

In [11]:
utheta=np.sqrt(u_atheta**2+u_btheta**2)

In [12]:
theta=ufloat(theta,utheta)

In [13]:
theta

0.3614867769088922+/-0.00043043428502439194

$$d \sin \theta=k \lambda$$

In [14]:
k=2
lam=ureg("589.3nm")
d=k*lam/umath.sin(theta)

In [15]:
d

<Quantity(3332.53+/-3.79, 'nanometer')>

$$u(d)=\frac{k \lambda \cos \theta}{\sin^2 \theta} u(\theta)$$

上面的结果是自动微分算出的，与使用公式的结果相同

In [16]:
ud=(2*lam*np.cos(theta.n)/np.sin(theta.n)**2*theta.s).magnitude
ud

np.float64(3.7937852279431414)

In [17]:
d.s

3.793785227943141

In [18]:
np.isclose(ud,d.s)

np.True_

所以$d=3333 \pm 4 nanometer$

## 2. 应用不等精度加权平均法计算氢原子的里德伯常数的最佳值 ，并与标准值比较，得出相对偏差。

In [19]:
氢灯折射角=pd.DataFrame({
    "k=1":(氢灯角度["k=-1 左"]+氢灯角度["k=-1 右"]-氢灯角度["k=+1 左"]-氢灯角度["k=+1 右"])/4,
    "k=2":(氢灯角度["k=-2 左"]+氢灯角度["k=-2 右"]-氢灯角度["k=+2 左"]-氢灯角度["k=+2 右"])/4,
    "k=3":(氢灯角度["k=-3 左"]+氢灯角度["k=-3 右"]-氢灯角度["k=+3 左"]-氢灯角度["k=+3 右"])/4,
})

In [20]:
氢灯折射角.index=[3,4,5]

In [21]:
氢灯折射角

,k=1,k=2,k=3
3,11.5167 degree,-66.6625 degree,-53.6000 degree
4,8.5292 degree,17.1000 degree,-63.8708 degree
5,7.4750 degree,nan degree,nan degree


In [22]:
for i in 氢灯折射角:
    氢灯折射角[i]=氢灯折射角[i].apply(lambda x: x.to('rad').magnitude)
    氢灯折射角[i]=氢灯折射角[i].apply(lambda x: x+np.pi/2 if x < 0 else x)

In [23]:
氢灯折射角

,k=1,k=2,k=3
3,0.2010,0.4073,0.6353
4,0.1489,0.2985,0.4560
5,0.1305,NaN,NaN


$$d \sin \theta=k \lambda$$

$$\frac{1}{\lambda}=R_H \left( \frac{1}{2^2}-\frac{1}{n^2} \right)$$

由上两式

$$R_H=\frac{k}{d \sin{\theta} \times (\frac{1}{2^2}-\frac{1}{n^2})}$$

为简便，记$f(d,\theta)=\frac{1}{d \sin{\theta}}$

$$u(R_H)=\frac{k}{\frac{1}{2^2}-\frac{1}{n^2}} \sqrt{ \left( \frac{\partial f}{\partial d} \right)^2 u^2(d) + \left( \frac{\partial f}{\partial \theta} \right)^2 u^2(\theta)}$$

$$\frac{\partial f}{\partial d}=-\frac{1}{d^2 \sin \theta}$$
$$\frac{\partial f}{\partial \theta}=-\frac{\cos \theta}{d \sin^2 \theta}$$

由于每组只测量了一次，所以

$$u(\theta)=u_b(\theta)$$

后续的数值计算是通过自动微分实现的，应当与公式结果一致

In [24]:
def 计算里德伯常数(k,n,theta):
    theta=ufloat(theta,u_btheta)
    lam=d*umath.sin(theta)/k
    RH=1/lam/(2**-2-n**-2)
    return RH

In [25]:
里德伯常数=[]
for i in 氢灯折射角:
    for j in range(3,6):
        result=计算里德伯常数(int(i[-1]),j,氢灯折射角[i][j])
        if not np.isnan(result.n):
            里德伯常数.append(result)

In [26]:
里德伯常数

[<Quantity(0.0108214+/-0.0000152, '1 / nanometer')>,
 <Quantity(0.0107906+/-0.0000172, '1 / nanometer')>,
 <Quantity(0.0109838+/-0.0000188, '1 / nanometer')>,
 <Quantity(0.0109077+/-0.0000131, '1 / nanometer')>,
 <Quantity(0.0108855+/-0.0000137, '1 / nanometer')>,
 <Quantity(0.0109224+/-0.0000127, '1 / nanometer')>,
 <Quantity(0.0109019+/-0.0000130, '1 / nanometer')>]

In [27]:
R_H最佳观测量=sum(i.n/i.s**2 for i in 里德伯常数)/sum(1/i.s**2 for i in 里德伯常数)

In [28]:
R_H最佳观测量

0.010889547041153715

In [29]:
uR_H=np.sqrt(1/sum(1/i.s**2 for i in 里德伯常数))

In [30]:
uR_H

np.float64(5.440136310889163e-06)

In [31]:
R_H=ufloat(R_H最佳观测量,uR_H)/ureg("nanometer")

In [32]:
R_H.to("1/m")

<Quantity((1.08895+/-0.00054)e+07, '1 / meter')>

测量得到里德伯常数为$(1.0890 \pm 0.0005) \times 10^7 \  {meter}^{-1}$

In [33]:
from scipy.constants import physical_constants

# 获取里德伯常数的详细信息
value, unit, uncertainty = physical_constants['Rydberg constant']

R_H_0=quan(value,unit)

In [34]:
R_H_0

<Quantity(10973731.568157, '1 / meter')>

In [35]:
相对偏差=((R_H最佳观测量/ureg("nanometer")-R_H_0)/R_H_0).to("").magnitude
print(f"{相对偏差=:.3%}")

相对偏差=-0.767%


相对偏差$=-0.767\%$

## 3. 计算钠黄光 k = 2 级谱线的角色散率和分辨本领.

角色散率
$$D_{\theta} = \frac{\delta \theta}{\delta \lambda} = \frac{k}{d \cos \theta}$$

In [36]:
D_theta = 2/d/umath.cos(theta)

In [37]:
D_theta.to("1/m")

<Quantity(641611+/-835, '1 / meter')>

角色散率
$$D_{\theta} = (6.416 \pm 0.008) \times 10^5 meter^{-1}$$

色分辨本领
$$R=\frac{\lambda}{\delta \lambda}=k N=k \frac{D}{d}$$

In [38]:
D=quan(2.20,"cm")
R=D/d*k

In [39]:
R.to("")

<Quantity(13203.2+/-15.0, 'dimensionless')>

色分辨本领
$$R=13203 \pm 15$$

## 4. 课后思考题

### 1. 本实验能否将钠黄光的双线分开，为什么？怎样才能把它们分开？

可以，实验过程中在$k=2$时确实观测到了两条分立的线状光谱
当$$\delta \lambda \geq \frac{\lambda}{kN}$$时可以分辨，等于时恰可分辨
钠灯的$\delta \lambda=0.6 nm$

In [40]:
delta_lam=quan(0.6,"nm")
k_inf=lam/delta_lam*d.n*d.u/D

In [41]:
k_inf.to("")

<Quantity(0.14877723374694982, 'dimensionless')>

计算表明$k=1$时就可以分辨钠的双黄线

### 2. 请查阅文献，列举光谱仪有哪些应用。

1. 原子结构研究：通过分析原子发射或吸收的特征谱线，反推原子的能级结构。

2. 物质成分鉴定：不同物质的光谱独一无二，可以用来进行定性与定量分析，在化学分析、环境监测中常用。

3. 天文学：分析恒星和星系的光谱，能知道它们的化学成分、温度、运动速度等。